File: JavaVersion/AoC2020/src/com/company/Day7.java
package com.company;

import java.util.ArrayList;

public class Day7 extends DefaultClass {

    ArrayList<Rule> Rules = new ArrayList<>();
    ArrayList<String> LookFor = new ArrayList<>();
    ArrayList<String> NextLookFor = new ArrayList<>();
    public Day7()
    {
        super(7,true);
        ArrayList<String> inp = fr.ReadLine();
        for (int i = 0; i < inp.size(); i++) {
            Rules.add(new Rule(inp.get(i)));
        }
        System.out.println("File Read and Translate: ");
        TimeResetClock();
        Part1();
        TimeResetClock();
        Part2();
        TimeClock();
    }

    private void Part1()
    {
        LookFor.add("shinygold");
        int sum =0;
        while(!LookFor.isEmpty())
        {
            for (int i = 0; i < Rules.size(); i++) {
                if(!Rules.get(i).Added)
                {
                    sum+=RuleContainBag(i);
                }
            }
            LookFor = new ArrayList<>(NextLookFor);
            NextLookFor.clear();
        }
        System.out.println("Part1: "+ sum);
    }

    private int RuleContainBag(int i)
    {
        for (int j = 0; j < Rules.get(i).ContainBag.size(); j++) {
            for (int k = 0; k < LookFor.size(); k++) {
                if(Rules.get(i).ContainBag.get(j).equals(LookFor.get(k)))
                {
                    NextLookFor.add(Rules.get(i).OutBag);
                    Rules.get(i).Added=true;
                    return 1;
                }
            }
        }
        return 0;
    }

    private void Part2()
    {
        int sum = NumberBags("shinygold");
        System.out.println("Part2:" + sum);
    }

    private int NumberBags(String BagName)
    {
        int sum = 0;
        for (int i = 0; i < Rules.size(); i++) {
            if(Rules.get(i).OutBag.equals(BagName))
            {
                if(Rules.get(i).ContainBag.size()==0)
                    return 0;
                for (int j = 0; j < Rules.get(i).ContainBag.size(); j++) {
                    sum += Rules.get(i).NumContainBag.get(j) * (NumberBags(Rules.get(i).ContainBag.get(j))+1);
                }
            }
        }
        return sum;
    }

    private class Rule
    {
        public String OutBag;
        public ArrayList<String> ContainBag = new ArrayList<>();
        public ArrayList<Integer> NumContainBag = new ArrayList<>();
        public boolean Added = false;

        public Rule(String str)
        {
            ArrayList<String> Split = fr.SplitString(str," ");
            OutBag = Split.get(0);
            OutBag += Split.get(1);
            for (int i = 4; i < Split.size(); i++) {
                try {
                    NumContainBag.add(Integer.parseInt(Split.get(i)));
                    i++;
                    ContainBag.add(Split.get(i));
                    i++;
                    ContainBag.set(ContainBag.size() - 1, ContainBag.get(ContainBag.size() - 1) + Split.get(i));
                    i++;
                }
                catch (Exception e)
                {
                    i=Split.size()+1;
                }
            }
        }
    }
}

File: CPPVersion/AoC2020CPP/AoC2020CPP/Day7.cpp
#include "Day7.h"

Day7::Day7() :DefaultClass(7, true)
{
	std::vector<std::string> inp = FR.ReadLine();
	for (int i = 0; i < inp.size(); i++)
	{
		CipherLine(inp[i]);
	}

	ResetClock();
	Part1();
	TimeResetClock();
	Part2();
	TimeClock();
}

void Day7::CipherLine(std::string Line)
{
	std::vector<std::string> Split = FR.SplitString(Line, ' ');
	OuterBag.push_back(BagLookUp(Split[0] + Split[1]));
	NumBagsContains.push_back(std::vector<int>());
	ContainBags.push_back(std::vector<int>());
	Visited.push_back(false);
	if (Split[4] == "no")
		return;
	for (int i = 4; i < Split.size(); i += 4)
	{
		NumBagsContains.at(NumBagsContains.size() - 1).push_back(std::stoi(Split[i]));
		ContainBags.at(ContainBags.size() - 1).push_back(BagLookUp(Split[i + 1] + Split[i + 2]));
	}
}

int Day7::BagLookUp(std::string colour)
{
	for (int i = 0; i < LookUp.size(); i++)
	{
		if (colour == LookUp[i])
			return i;
	}
	LookUp.push_back(colour);
	return LookUp.size() - 1;
}
void Day7::Part1()
{
	std::vector<int> Lookfor;
	std::vector<int> NextLookfor;
	Lookfor.push_back(BagLookUp("shinygold"));
	int sum = 0;
	while (!Lookfor.empty())
	{
		for (int i = 0; i < ContainBags.size(); i++)
		{
			if (!Visited.at(i))
			{
				if (IsContained(&ContainBags.at(i), &Lookfor))
				{
					NextLookfor.push_back(OuterBag.at(i));
					Visited.at(i) = true;
					sum++;
				}
			}
		}
		Lookfor = NextLookfor;
		NextLookfor.clear();
	}
	std::cout << "Part1: " << sum<<"\n";
}

bool Day7::IsContained(std::vector<int>* Contained, std::vector<int>* Lookfor)
{
	for (int i = 0;  i < Contained->size();  i++)
	{
		for (int j = 0; j < Lookfor->size(); j++)
		{
			if (Contained->at(i) == Lookfor->at(j))
				return true;
		}
	}
	return false;
}

void Day7::Part2()
{
	int x = BagLookUp("shinygold");
	std::cout << "Part2: " << NumBags(x)<<"\n";
}

int Day7::NumBags(int pos)
{
	int sum = 0;
	for (int i = 0; i < OuterBag.size(); i++)
	{
		if (OuterBag.at(i) == pos)
		{
			if (ContainBags.at(i).empty())
				return 0;
			else
			{
				for (int j = 0; j < ContainBags.at(i).size(); j++)
				{
					sum += (NumBagsContains.at(i).at(j) * (NumBags(ContainBags.at(i).at(j)) + 1));
				}
			}
		}
	}
	return sum;
}
